File: Leetcode/53. Maximum Subarray/53_Maximum_Subarray.py
from typing import List


class Solution:
    def masSubArray(self, nums: List[int]) -> List[int]:
        result = nums[0]
        total = 0
        for n in nums:
            total += n
            result = max(result, total)
            if total < 0:
                total = 0
        return result


class Main:
    def main(self):
        nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
        solution = Solution()
        result = solution.masSubArray(nums)
        print(result)


if __name__ == "__main__":
    Main().main()

File: Leetcode/53. Maximum Subarray/53_Maximum_Subarray.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int maxSubArray(vector<int> &nums)
    {
        int current = nums[0];
        int result = nums[0];
        for (int i = 1; i < nums.size(); i++)
        {
            current = max(current + nums[i], nums[i]);
            result = max(current, result);
        }
        return result;
    }
};

int main()
{
    vector<int> nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    Solution s;
    int result = s.maxSubArray(nums);
    cout << result << endl;
}